atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/watch/LockWatchEventLogImpl.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.api.watch;
+
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.Streams;
+import com.palantir.lock.watch.CacheStatus;
+import com.palantir.lock.watch.IdentifiedVersion;
+import com.palantir.lock.watch.LockWatchCreatedEvent;
+import com.palantir.lock.watch.LockWatchEvent;
+import com.palantir.lock.watch.LockWatchStateUpdate;
+import com.palantir.logsafe.Preconditions;
+
+final class LockWatchEventLogImpl implements LockWatchEventLog {
+    private final ClientLockWatchSnapshot snapshot;
+    private final VersionedEventStore eventStore = new VersionedEventStore();
+
+    private Optional<IdentifiedVersion> latestVersion = Optional.empty();
+
+    static LockWatchEventLog create() {
+        return create(ClientLockWatchSnapshotImpl.create());
+    }
+
+    @VisibleForTesting
+    static LockWatchEventLog create(ClientLockWatchSnapshot snapshot) {
+        return new LockWatchEventLogImpl(snapshot);
+    }
+
+    private LockWatchEventLogImpl(ClientLockWatchSnapshot snapshot) {
+        this.snapshot = snapshot;
+    }
+
+    @Override
+    public CacheStatus processUpdate(LockWatchStateUpdate update) {
+        if (!latestVersion.isPresent() || !update.logId().equals(latestVersion.get().id())) {
+            return update.accept(new NewLeaderVisitor());
+        } else {
+            return update.accept(new ProcessingVisitor());
+        }
+    }
+
+    @Override
+    public ClientLogEvents getEventsBetweenVersions(
+            Optional<IdentifiedVersion> lastKnownVersion,
+            IdentifiedVersion endVersion) {
+        Optional<IdentifiedVersion> startVersion = lastKnownVersion.map(this::createStartVersion);
+        IdentifiedVersion currentVersion = getLatestVersionAndVerify(endVersion);
+        ClientLogEvents.Builder eventBuilder = new ClientLogEvents.Builder();
+        final long fromSequence;
```

Comment:
Even better: I would just have an if statement that run the whole codepath, it is clearly a different codepath depending on whether you have a snapshot or not. The way you have it you're saving 1 line of duplication (to do the getEventsBetweenVersionsInclusive in one place), but in order to do that you are making the code non-linear and creating a builder half way through. My opinion is that this would be easier to read if you just had:

```
if (rebuildFromSnapshot()) {
  return new ClientLogEvents.Builder()
    .clearCache(true)
    .addEvents(LockWatchCreatedEvent.fromSnapshot(snapshot.getSnapshot()))
    .addAllEvents(eventStore.getEventsBetweenVersionsInclusive(Optional.empty(), endVersion.version()))
}

return new ClientLogEvents.Builder()
    .clearCache(false)
    .addAllEvents(eventStore.getEventsBetweenVersionsInclusive(Optional.of(theVersion), endVersion.version()));
```

Reply:
Done, although I feel like there is maybe a nicer optional functional method to use in my impl to make it nicer.

Reply:
ah, didn't see later comment.

Reply:
Actually that does work out pretty nicely.